test/src/unit-cppapi-consolidation-with-timestamps.cc

```diff
@@ -0,0 +1,449 @@
+  return vfs_.is_dir(array_name);
+}
+
+TEST_CASE_METHOD(
+    ConsolidationWithTimestampsFx,
+    "CPP API: Test consolidation with timestamps",
+    "[cppapi][consolidation-with-timestamps][write-check]") {
+  remove_sparse_array();
+  create_sparse_array();
+
+  // Write first fragment.
+  write_sparse(
+      {0, 1, 2, 3, 4, 5, 6, 7},
+      {1, 1, 1, 2, 3, 4, 3, 3},
+      {1, 2, 4, 3, 1, 2, 3, 4},
+      1);
+
+  // Write second fragment.
+  write_sparse({8, 9, 10, 11}, {2, 2, 3, 3}, {2, 3, 2, 3}, 2);
+
+  // Consolidate.
+  consolidate_sparse();
+
+  // Check t.tdb file.
+  const uint64_t X = std::numeric_limits<uint64_t>::max();
+  check_timestamps_file({1, 1, 2, 1, X, X, 1, 2, 1, X, X, 1});
+
+  remove_sparse_array();
+}
+
+TEST_CASE_METHOD(
+    ConsolidationWithTimestampsFx,
+    "CPP API: Test consolidation with timestamps, global read",
+    "[cppapi][consolidation-with-timestamps][global-read]") {
+  remove_sparse_array();
+  create_sparse_array();
+
+  // Write first fragment.
+  write_sparse(
+      {0, 1, 2, 3, 4, 5, 6, 7},
+      {1, 1, 1, 2, 3, 4, 3, 3},
+      {1, 2, 4, 3, 1, 2, 3, 4},
+      1);
+
+  // Write second fragment.
+  write_sparse({8, 9, 10, 11}, {2, 2, 3, 3}, {2, 3, 2, 3}, 2);
+
+  // Consolidate.
+  consolidate_sparse();
+
+  std::string stats;
+  std::vector<int> a1(10);
+  std::vector<uint64_t> dim1(10);
+  std::vector<uint64_t> dim2(10);
+  read_sparse_global_order(a1, dim1, dim2, stats);
+
+  std::vector<int> c_a1 = {0, 1, 8, 2, 9, 4, 10, 5, 11, 7};
+  std::vector<uint64_t> c_dim1 = {1, 1, 2, 1, 2, 3, 3, 4, 3, 3};
+  std::vector<uint64_t> c_dim2 = {1, 2, 2, 4, 3, 1, 2, 2, 3, 4};
+  CHECK(!memcmp(c_a1.data(), a1.data(), sizeof(c_a1)));
+  CHECK(!memcmp(c_dim1.data(), dim1.data(), sizeof(c_dim1)));
+  CHECK(!memcmp(c_dim2.data(), dim2.data(), sizeof(c_dim2)));
+
+  remove_sparse_array();
+}
+
+TEST_CASE_METHOD(
+    ConsolidationWithTimestampsFx,
+    "CPP API: Test consolidation with timestamps, global read, all cells same "
+    "coords",
+    "[cppapi][consolidation-with-timestamps][global-read][same-coords]") {
+  remove_sparse_array();
+  create_sparse_array();
+
+  // Write fragments.
+  for (uint64_t i = 0; i < 50; i++) {
+    std::vector<int> a(1);
+    a[0] = i + 1;
+    write_sparse(a, {1}, {1}, i + 1);
+  }
+
+  // Consolidate.
+  consolidate_sparse();
+
+  std::string stats;
+  std::vector<int> a1(1);
+  std::vector<uint64_t> dim1(1);
+  std::vector<uint64_t> dim2(1);
+  read_sparse_global_order(a1, dim1, dim2, stats);
+
+  CHECK(a1[0] == 50);
+  CHECK(dim1[0] == 1);
+  CHECK(dim2[0] == 1);
+
+  remove_sparse_array();
+}
+
+TEST_CASE_METHOD(
+    ConsolidationWithTimestampsFx,
+    "CPP API: Test consolidation with timestamps, global read, same cells "
+    "across tiles",
+    "[cppapi][consolidation-with-timestamps][global-read][across-tiles]") {
+  remove_sparse_array();
+  create_sparse_array();
+
+  // Write fragments.
+  for (uint64_t i = 0; i < 6; i++) {
```

Comment:
For my undestanding, can you explain why those writes result in same cells across result tiles? I don't manage to understand how this works :(

Reply:
Yes, we write 8 cells per fragments for 6 fragments. Then it gets consolidated into one. So you'll get in order 6xcell1, 6xcell2... 48 cells. Tile capacity is 20 so you'll end up with 3 tiles. First break in the tiles will be in the middle of cell3, second will be in the middle of the cells7. Does that make sense?

Reply:
Thanks! Makes perfect sense, so perfect that it would be a great as a comment in this test 😄  😄